sdk/scheduler/src/main/java/com/mesosphere/sdk/dcos/auth/CachedTokenProvider.java

```diff
@@ -0,0 +1,69 @@
+package com.mesosphere.sdk.dcos.auth;
+
+import com.auth0.jwt.interfaces.DecodedJWT;
+
+import java.io.IOException;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.Optional;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+/**
+ * CachedTokenProvider retrieves token from underlying provider and caches the value. It automatically triggers
+ * getToken() method on underlying provider when token is about to expire.
+ */
+public class CachedTokenProvider implements TokenProvider {
+
+    private final TokenProvider provider;
+    private Optional<DecodedJWT> token;
+    private final Duration ttl;
+
+    private final ReadWriteLock internalLock = new ReentrantReadWriteLock();
+    private final Lock rlock = internalLock.readLock();
+    private final Lock rwlock = internalLock.writeLock();
+
+
+    public CachedTokenProvider(TokenProvider provider, Duration ttl) {
+        this.provider = provider;
+        this.ttl = ttl;
+        this.token = Optional.empty();
+    }
+
+    public CachedTokenProvider(TokenProvider provider) {
+        this(provider, Duration.ofSeconds(30));
```

Comment:
Could remove this version of the constructor, now that the default is being handled upstream in SchedulerFlags